inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/special/ScheduledExecutorContextPropagationSensor.java

```diff
@@ -0,0 +1,82 @@
+package rocks.inspectit.oce.core.instrumentation.special;
+
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.asm.AsmVisitorWrapper;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.dynamic.DynamicType;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.springframework.context.annotation.DependsOn;
+import org.springframework.stereotype.Component;
+import rocks.inspectit.oce.bootstrap.Instances;
+import rocks.inspectit.oce.core.instrumentation.config.model.InstrumentationConfiguration;
+import rocks.inspectit.oce.core.instrumentation.context.ContextManagerImpl;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static net.bytebuddy.matcher.ElementMatchers.*;
+
+/**
+ * Special sensor for passing the context via the {@link ScheduledExecutorService}.
+ * This sensor will pass the context when using the following methods:
+ * <p><ul>
+ * <li>{@link ScheduledExecutorService#schedule(Runnable, long, TimeUnit)}
+ * <li>{@link ScheduledExecutorService#schedule(Callable, long, TimeUnit)}
+ * <li>{@link ScheduledExecutorService#scheduleWithFixedDelay(Runnable, long, long, TimeUnit)}
+ * <li>{@link ScheduledExecutorService#scheduleAtFixedRate(Runnable, long, long, TimeUnit)}
+ * </ul><p>
+ */
+@Component
+@DependsOn(ContextManagerImpl.BEAN_NAME)
+public class ScheduledExecutorContextPropagationSensor implements SpecialSensor {
+
+    private static final ElementMatcher<TypeDescription> CLASSES_MATCHER = isSubTypeOf(ScheduledExecutorService.class);
+
+    @Override
+    public boolean shouldInstrument(TypeDescription type, InstrumentationConfiguration settings) {
+        return settings.getSource().getSpecial().isScheduledExecutorContextPropagation() && CLASSES_MATCHER.matches(type);
+    }
+
+    @Override
+    public boolean requiresInstrumentationChange(TypeDescription type, InstrumentationConfiguration first, InstrumentationConfiguration second) {
+        return false; //if the sensor stays active it never requires changes
+    }
+
+    @Override
+    public DynamicType.Builder instrument(Class<?> clazz, TypeDescription type, InstrumentationConfiguration settings, DynamicType.Builder builder) {
+        return builder
+                .visit(ScheduledExecutorRunnableAdvice.TARGET)
+                .visit(ScheduledExecutorCallableAdvice.TARGET);
+    }
+
+    /**
+     * Advice for wrapping the first method argument - which has to be a {@link Runnable} - into a Runnable for attaching and detaching the current context.
+     * See also {@link io.grpc.Context#wrap(Runnable)}
+     */
+    private static class ScheduledExecutorRunnableAdvice {
+
+        static final AsmVisitorWrapper.ForDeclaredMethods TARGET = Advice.to(ScheduledExecutorRunnableAdvice.class)
+                .on(nameStartsWith("schedule").and(takesArgument(0, Runnable.class)));

```

Comment:
I was thinking if we need to somehow instrument this more smarter, but I could not find any solution.. The shitty thing is that if you take for example a `java.util.concurrent.Executors.DelegatedScheduledExecutorService` we would actually wrap the callable/runnable twice.. However, I see no way in not doing this, we could alternatively add that class is not abstract as in the Executor instrumentation, at least we would then only instrument concrete classes that implement the interface and implement the method.. We would decrease the amount of instrumentation maybe for a bit maybe..

You can also add a system-test with `DelegatedScheduledExecutorService`.

Reply:
Yes, added the matcher that it should not be abstract.